controllers/perconaservermysqlbackup_controller.go

```diff
@@ -45,16 +67,220 @@ type PerconaServerMySQLBackupReconciler struct {
+			exists, err := k8s.ObjectExists(ctx, r.Client, nn, &corev1.Secret{})
+			if err != nil {
+				return rr, errors.Wrapf(err, "check %s exists", nn)
+			}
+
+			if !exists {
+				return rr, errors.Errorf("secret %s not found", nn)
+			}
+
+			if err := xtrabackup.SetStorageGCS(job, storage.GCS); err != nil {
+				return rr, errors.Wrap(err, "set storage GCS")
+			}
+		case apiv1alpha1.BackupStorageAzure:
+			if storage.Azure == nil {
+				return rr, errors.New("azure stanza is required in storage")
+			}
+
+			nn := types.NamespacedName{Name: storage.Azure.CredentialsSecret, Namespace: cr.Namespace}
+			exists, err := k8s.ObjectExists(ctx, r.Client, nn, &corev1.Secret{})
+			if err != nil {
+				return rr, errors.Wrapf(err, "check %s exists", nn)
+			}
+
+			if !exists {
+				return rr, errors.Errorf("secret %s not found", nn)
+			}
+
+			if err := xtrabackup.SetStorageAzure(job, storage.Azure); err != nil {
+				return rr, errors.Wrap(err, "set storage Azure")
+			}
+		default:
+			return rr, errors.Errorf("storage type %s is not supported", storage.Type)
+		}
+
+		src, err := r.getBackupSource(ctx, cluster)
+		if err != nil {
+			return rr, errors.Wrap(err, "get backup source node")
+		}
+
+		if err := xtrabackup.SetSourceNode(job, src); err != nil {
+			return rr, errors.Wrap(err, "set backup source node")
+		}
+
+		if err := controllerutil.SetControllerReference(cr, job, r.Scheme); err != nil {
+			return rr, errors.Wrapf(err, "set controller reference to Job %s/%s", job.Namespace, job.Name)
+		}
+
+		if err := r.Client.Create(ctx, job); err != nil {
+			return rr, errors.Wrapf(err, "create job %s/%s", job.Namespace, job.Name)
+		}
+
+		return rr, nil
+	}
+
+	switch status.State {
+	case apiv1alpha1.BackupStarting:
+		if j.Status.Active > 0 {
+			status.State = apiv1alpha1.BackupRunning
+		}
+	case apiv1alpha1.BackupRunning:
+		if j.Status.Active > 0 {
+			return rr, nil
+		}
+
+		for _, cond := range j.Status.Conditions {
+			if cond.Status != corev1.ConditionTrue {
+				continue
+			}
+
+			switch cond.Type {
+			case batchv1.JobFailed:
+				status.State = apiv1alpha1.BackupFailed
+			case batchv1.JobComplete:
+				status.State = apiv1alpha1.BackupSucceeded
+			}
+		}
+	case apiv1alpha1.BackupFailed, apiv1alpha1.BackupSucceeded:
+		return rr, nil
+	default:
+		status.State = apiv1alpha1.BackupStarting
+	}
+
+	return rr, nil
 }
 
-// SetupWithManager sets up the controller with the Manager.
-func (r *PerconaServerMySQLBackupReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&apiv1alpha1.PerconaServerMySQLBackup{}).
-		Complete(r)
+func (r *PerconaServerMySQLBackupReconciler) getBackupSource(ctx context.Context, cluster *apiv1alpha1.PerconaServerMySQL) (string, error) {
+	l := log.FromContext(ctx).WithName("getBackupSource")
+
+	orcHost := orchestrator.APIHost(cluster)
+	primary, err := orchestrator.ClusterPrimary(ctx, orcHost, cluster.ClusterHint())
```

Comment:
Can we avoid using  orchestrator here? Because if something is wrong with orchestrator our db cluster can't correctly work at all. We always need to think about it.

Reply:
Also what about the setup with group replication which will not have orchestrator, if we use it somewhere we need to make it conditional everywhere.

Reply:
It will be improved via separate task. @egegunes please create it.

Reply:
https://jira.percona.com/browse/K8SPS-87